### PR TITLE
Introduce usage of DOAP file

### PR DESCRIPTION
**Pull request by melvo**
PR check list:
- [x] Document your code
- [x] Add `\since QXmpp 1.X`
- [x] Fix doxygen warnings (see log when building with `-DBUILD_DOCUMENTATION=ON`)
- [x] Update `doc/xep.doc`
- [x] Add unit tests
- [x] Format the code: Run `clang-format -i src/<edited-file(s)> tests/<edited-file(s)>`

### CODE REVIEW

**Comment by melvo** [comment]
@lnjX The DOAP file can be used to [create a human-readable page](/pulkomandy/xmpp-doap). QXmpp's [Doxygen page](https://doc.qxmpp.org/qxmpp-1/pages.html) could link to it. We are working on such a page located on https://xmpp.org/software/libraries/.

**Comment by lnjX** [comment]
It's nice that it's possible to generate such a page. What I'm still missing to continue here is a non-external replacement for xep.doc that is generated from the DOAP file.

**Comment by melvo** [comment]
@lnjX The XEP list is now integrated in the Doxygen documentation.

File: doc/doap-iframe.html [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<html>
+	<head>
+		<script>
+			function resizeIframe(obj) {
+				obj.style.height = obj.contentWindow.document.documentElement.scrollHeight + 20 + 'px';
+			}
+		</script>
```
Comment (lnjX) [inline doc/doap-iframe.html:7]: please use css for styling

File: doc/xeplist.xml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...on><sig>Standards JIG</sig></xep><xep accepted="true"><number>233</number><title>XMPP Server Registration for use with Kerberos V5</title><abstract>This specification defines the Kerberos principal name of an XMPP server. It also details a method by which a connecting client can determine this Kerberos principal name when authenticating using the "GSSAPI" SASL mechanism.</abstract><type>Standards Track</type><status>Draft</status><approver>Council</approver><shortname>kerberos5</shortname><last-revision><date>2017-03-16</date><version>1.0.0</version><initials>XEP Editor (ssw)</initials><remark>Move to draft by order of the XMPP Council.</remark></last-revision><sig>Standards</sig><lastcall>2017-03-01</lastcall></xep><xep accepted="true"><number>257</number><title>Client Certificate Management for SASL EXTERNAL</title><abstract>This specification defines a method to manage client certificates that can be used with SASL External to allow clients to log in without a password.</abstract><type>Standards Track</type><status>Deferred</status><approver>Council</approver><last-revision><date>2012-07-18</date><version>0.3</version><initials>tpa</initials><remark>* Removed the dependency on .
```
Comment (lnjX) [inline doc/xeplist.xml:1]: how does updating of this file work?

File: doc/doap-iframe.html (hunk 1 above)
Comment (melvo) [reply doc/doap-iframe.html]: Unfortunately, that is not possible. I added a corresponding comment explaining it to that function.

File: doc/xeplist.xml (hunk 2 above)
Comment (melvo) [reply doc/xeplist.xml]: The latest version is fetched now before generating the documentation.

File: doc/CMakeLists.txt [hunk 3]
```diff
@@ -16,7 +16,9 @@ configure_file(
 
 add_custom_target(
     doc ALL
+    COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/fetch-xeplist.sh
     COMMAND ${DOXYGEN_EXECUTABLE} ${CMAKE_CURRENT_BINARY_DIR}/Doxyfile
+    COMMAND rm ${CMAKE_CURRENT_SOURCE_DIR}/xeplist.xml
     WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
     COMMENT "Generate API documentation with Doxygen"
     VERBATIM
```
Comment (lnjX) [inline doc/CMakeLists.txt:24]: * have you tested this on windows?
* this way it's not possible to generate the documentation when you're offline, right? debian requires that building works offline (and I think that's a good idea)

Comment (melvo) [reply doc/CMakeLists.txt]: The XEP list is now a local copy in QXmpp's repository and can be updated by `doc/update-xeplist.sh`.

File: doc/doap.css [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    text-align: end;
+}
+a {
+	padding: 5px;
+	color: rgb(70, 130, 180);
+    text-decoration: none;
+}
+a:hover {
+	color: rgb(50, 90, 120);
+    text-decoration: none;
```
Comment (lnjX) [inline doc/doap.css:51]: can you use 4 spaces as indentation in the whole file?

File: doc/doap.xsl [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+			<span class="{xmpp:status}"><xsl:value-of select="xmpp:status"/></span>
+		</td>
+		<td>
+			<span class="version"><xsl:value-of select="xmpp:since"/></span>
+		</td>
+		<td class="small">
+			<xsl:value-of select="xmpp:note"/>
+		</td>
+	</tr>
+</xsl:template>
```
Comment (lnjX) [inline doc/doap.xsl:86]: can we use 4 spaces here too?

File: doc/doap.css (hunk 4 above)
Comment (melvo) [reply doc/doap.css:51]: It is the indentation of the original file. The next step will be a mechanism to update the files coming from https://github.com/pulkomandy/xmpp-doap. Thus, I would not modify the files more than needed.

File: doc/doap.xsl (hunk 5 above)
Comment (melvo) [reply doc/doap.xsl:86]: See https://github.com/qxmpp-project/qxmpp/pull/438#discussion_r961674485

File: doc/doap.css (hunk 4 above)
Comment (melvo) [reply doc/doap.css:51]: But it seems that the indentation has changed. I will get the latest version.

File: doc/doap.xml [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+SPDX-License-Identifier: CC0-1.0
+-->
+
+<rdf:RDF xmlns:rdf='http://www.w3.org/1999/02/22-rdf-syntax-ns#'
+         xmlns='http://usefulinc.com/ns/doap#'
+         xmlns:xmpp='https://linkmauve.fr/ns/xmpp-doap#'
+         xmlns:schema='https://schema.org/'>
+  <Project>
+    <name>QXmpp</name>
+    <created>2009-01-25</created>
```
Comment (lnjX) [inline doc/doap.xml:16]:
```suggestion
    <created>2008-12-01</created>
```
as far as I know QXmpp was already started in dec 2008, in 2009 the svn repo was created

Comment (melvo) [reply doc/doap.xml]: Do you have any reference for the exact date?

Comment (lnjX) [reply doc/doap.xml]: unfortunately not

File: doc/doap.css (hunk 4 above)
Comment (lnjX) [reply doc/doap.css:51]: okay, so if the indentation is still mixed we should also fix that upstream

Comment (melvo) [reply doc/doap.css:51]: It is already fixed upstream. I will push the change.

File: doc/doap.xml (hunk 6 above)
Comment (melvo) [reply doc/doap.xml]: Where do you have that information from?

Comment (lnjX) [reply doc/doap.xml]: 2008-20xx was used in the copyright headers, I don't know where I read about december (but december 01 is probably a good estimate)

Comment (melvo) [reply doc/doap.xml]: OK.

**Comment by lnjX** [comment]
image
I ran `make doc` and opened the file locally with firefox and I got this.

**Comment by lnjX** [comment]
```
$ reuse lint
# MISSING COPYRIGHT AND LICENSING INFORMATION

The following files have no copyright and licensing information:
* ../SECURITY.md
* ../doc/doap-iframe.html
* ../doc/doap.css
* ../doc/doap.xsl
* ../doc/xeplist.xml
* ../src/omemo/README.md


# SUMMARY

* Bad licenses:
* Deprecated licenses:
* Licenses without file extension:
* Missing licenses:
* Unused licenses:
* Used licenses: CC0-1.0, LGPL-2.1-or-later, LicenseRef-MIT-variant
* Read errors: 0
* Files with copyright information: 399 / 413
* Files with license information: 399 / 413

Unfortunately, your project is not compliant with version 3.0 of the REUSE Specification :-(
```

please fix the missing license information in `doc/`

**Comment by melvo** [comment]
> image I ran `make doc` and opened the file locally with firefox and I got this.

You have to run a local server for that purpose because [Firefox prohibits the request](https://developer.mozilla.org/docs/Web/HTTP/CORS/Errors/CORSRequestNotHttp) to `doap.xsl`.
Try the following:
```
cd build/doc/html/
python3 -m http.server
firefox http://0.0.0.0:8000/xep.html
```

**Comment by melvo** [comment]
> please fix the missing license information in `doc/`

I created https://github.com/pulkomandy/xmpp-doap/issues/16 for that because the project itself does not provide any license information.

File: doc/doap-iframe.html [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+			<!-- Dynamic resizing of the frame is needed because its size inside of the page is unknown before rendering it. -->
+			function resizeIframe(obj) {
+				obj.style.height = obj.contentWindow.document.documentElement.scrollHeight + 20 + 'px';
+			}
+		</script>
+	</head>
+	<body>
+		<iframe src="doap.xml" class="xeps" width="100%" height="100%" frameborder="0" onload="resizeIframe(this)"/>
+	</body>
+</html>
```
Comment (lnjX) [inline doc/doap-iframe.html:19]: why have you readded the `<html/>` tag? having `<html/>` inside of the main `<html/>` is not allowed.

File: doc/update-xeplist.sh [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+# It should be run before generating the Doxygen documentation.
+#
+
+# The list is generated by "https://github.com/xsf/xeps/blob/master/tools/extract-metadata.py".
+SOURCE_URL="https://xmpp.org/extensions/xeplist.xml"
+
+QXMPP_DOC=$(dirname "$(readlink -f "${0}")")
+DESTINATION_PATH="${QXMPP_DOC}/xeplist.xml"
+
+curl -L ${SOURCE_URL} > ${DESTINATION_PATH}
```
Comment (lnjX) [inline doc/update-xeplist.sh:19]: overwrites the license header

File: doc/doap-iframe.html (hunk 7 above)
Comment (lnjX) [reply doc/doap-iframe.html]: I can't find it on github anymore. I'm pretty sure I did push that (but it's of course possible that I somehow forgot it).

File: .github/pull_request_template.md [hunk 9]
```diff
@@ -2,7 +2,7 @@ PR check list:
 - [ ] Document your code
 - [ ] Add `\since QXmpp 1.X`, `QXMPP_EXPORT`
 - [ ] Fix doxygen warnings (see log when building with `-DBUILD_DOCUMENTATION=ON`)
-- [ ] Update `doc/doap.xml` and run `update-xeplist.sh`
+- [ ] Update `doc/doap.xml`
```
Comment (melvo) [inline .github/pull_request_template.md:5]: @lnj We should document it somewhere else then.

File: doc/Doxyfile.in [hunk 10]
```diff
@@ -10,9 +10,9 @@ INPUT                  = @PROJECT_SOURCE_DIR@/doc/index.doc \
                          @CMAKE_BINARY_DIR@/src/base/QXmppBuildConstants.h \
                          @PROJECT_SOURCE_DIR@/src
 HTML_EXTRA_FILES       = @PROJECT_SOURCE_DIR@/doc/doap.xml \
-                         @PROJECT_SOURCE_DIR@/doc/doap.xsl \
-                         @PROJECT_SOURCE_DIR@/doc/xeplist.xml \
-                         @PROJECT_SOURCE_DIR@/doc/doap.css
+                         @PROJECT_SOURCE_DIR@/doc/doap-style/doap.xsl \
+                         @PROJECT_SOURCE_DIR@/doc/doap-style/xeplist.xml \
```
Comment (melvo) [inline doc/Doxyfile.in:14]: @lnjX `xeplist.xml` is no style. It only contains XEP metadata. Thus, the directory name should be adjusted accordingly.

Comment (lnjX) [reply doc/Doxyfile.in:14]: what's your suggestion?

File: doc/doap.xml [hunk 11]
```diff
@@ -0,0 +1,745 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<?xml-stylesheet href="doap.xsl" type="text/xsl"?>
```
Comment (melvo) [inline doc/doap.xml:2]: @lnj `href` must be adjusted too if you change the file's path.

File: doc/Doxyfile.in (hunk 10 above)
Comment (melvo) [reply doc/Doxyfile.in:14]: Maybe `doap`?

Comment (melvo) [reply doc/Doxyfile.in:14]: Your intention was to separate the content, which is changed frequently by developers, from the other parts, right?

Comment (lnjX) [reply doc/Doxyfile.in:14]: yes, especially to separate the stuff for generating the styled xep list

Comment (lnjX) [reply doc/Doxyfile.in:14]: the `doap` folder doesn't contain the doap.xml, that could be misleading

File: doc/doap.xml (hunk 11 above)
Comment (lnjX) [reply doc/doap.xml:2]: isn't `doap.xsl` also installed into the same location as `doap.xml`?

File: doc/Doxyfile.in (hunk 10 above)
Comment (melvo) [reply doc/Doxyfile.in:14]: I see. How about `doap-externals`?

File: doc/doap.xml (hunk 11 above)
Comment (melvo) [reply doc/doap.xml:2]: Ah, it is OK then.

File: doc/Doxyfile.in (hunk 10 above)
Comment (melvo) [reply doc/Doxyfile.in:14]: Or `doap-rendering`?

**Comment by melvo** [comment]
@lnjX I added missing information to the DOAP file and rebased everything.

File: doc/doap.xml [hunk 12]
```diff
@@ -0,0 +1,808 @@
...
+        <xmpp:since>0.5</xmpp:since>
+      </xmpp:SupportedXep>
+    </implements>
+    <implements>
+      <xmpp:SupportedXep>
+        <xmpp:xep rdf:resource='https://xmpp.org/extensions/xep-0060.html'/>
+        <xmpp:status>partial</xmpp:status>
+        <xmpp:version>1.24</xmpp:version>
+        <xmpp:since>0.3</xmpp:since>
+        <xmpp:note>Manager since 1.5</xmpp:note>
```
Comment (lnjX) [inline doc/doap.xml:126]: <1.5 only had an unfinished IQ class, I'd just mark it as since 1.5

File: doc/doap.xml [hunk 13]
```diff
@@ -0,0 +1,808 @@
...
+        <xmpp:since>0.4</xmpp:since>
+      </xmpp:SupportedXep>
+    </implements>
+    <implements>
+      <xmpp:SupportedXep>
+        <xmpp:xep rdf:resource='https://xmpp.org/extensions/xep-0198.html'/>
+        <xmpp:status>complete</xmpp:status>
+        <xmpp:version>1.5</xmpp:version>
+        <xmpp:since>1.0</xmpp:since>
+        <xmpp:note>No server-side support</xmpp:note>
```
Comment (lnjX) [inline doc/doap.xml:314]: that's the case for nearly all XEPs here

File: doc/doap.xml (hunk 12 above)
Comment (melvo) [reply doc/doap.xml:126]:
```suggestion
        <xmpp:since>1.5</xmpp:since>
```

File: doc/doap.xml (hunk 13 above)
Comment (melvo) [reply doc/doap.xml:313]:
```suggestion
```

**Comment by melvo** [comment]
@lnjX How can we temporarily update `xeplist.xml` in order to be used by https://doc.qxmpp.org/qxmpp-dev/?

**Comment by lnjX** [comment]
> @lnjX How can we temporarily update xeplist.xml in order to be used by https://doc.qxmpp.org/qxmpp-dev/?

You mean only updating it in the generated version, but not in the main qxmpp repository?

We can add it to the CI job.

**Review by lnjX — APPROVED** [review]

**Comment by melvo** [comment]
> > @lnjX How can we temporarily update xeplist.xml in order to be used by https://doc.qxmpp.org/qxmpp-dev/?
> 
> You mean only updating it in the generated version, but not in the main qxmpp repository?
> 
> We can add it to the CI job.

Yes.  That would be useful :)

**Pull request merged**